### ISSUE TITLE
Functions from fully qualified trait are resolved incorrectly

### ISSUE THREAD

**Issue by Kobzol** [intellij-rust/intellij-rust#5483]
## Environment
* **IntelliJ Rust plugin version:** 0.2.123.SNAPSHOT-201-dev
* **Rust toolchain version:** 1.43.1 (8d69840ab 2020-05-04) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2020.1.1 (CL-201.7223.86)
* **Operating system:** Linux 5.0.0-37-generic

## Problem description
Trait functions called with the fully qualified syntax `<Foo as Trait>::function` are resolved to `Foo::function` instead of `Trait::function`. This only happens if `Foo::function` exists, otherwise the resolution is correct.

## Steps to reproduce
```rust
use std::path::Path;
trait Trait {
    fn new(x: &str) -> i32;
}
impl Trait for Path {
    fn new(x: &str) -> i32 {
        123
    }
}
fn main() {
    <Path as Trait>::/*caret*/new("fo"); // resolved to Path::new, instead of Trait::new
}
```

**Comment by sollyucko** [comment]
I encountered this bug with two traits with the same method name. This even lead it to suggest that I import the other trait. In that case, the traits and implementations were all in the same file, separate from my code. The implementation of the correct trait was generated with `macro_rules!`. The implementation of the incorrect trait only used default implementations, and was a blanket `impl`.

Working repro:

```rust
use serde_hex::{SerHex, Strict};
use serde_json::Serializer;

fn foo() {
    <u8 as SerHex<Strict>>::serialize(&0_u8, Vec::<u8>::new())
}
```

[SerHex](https://docs.rs/serde-hex/0.1.0/serde_hex/trait.SerHex.html), [SerHexOpt](https://docs.rs/serde-hex/0.1.0/serde_hex/trait.SerHexOpt.html)

![Screenshot_2020-06-25_14-54-27](https://user-images.githubusercontent.com/15565481/85781623-e4ca6900-b6f3-11ea-80f1-c5332a3d0510.png)

---

Another case, that asks me to choose which declaration to jump to:

```rust
use std::path::Path;

trait Trait1 {
    fn foo(x: &str) -> i32;
}

trait Trait2 {
    fn foo(x: &str) -> i32;
}

impl Trait1 for Path {
    fn foo(x: &str) -> i32 {
        123
    }
}

impl Trait2 for Path {
    fn foo(x: &str) -> i32 {
        456
    }
}

fn main() {
    <Path as Trait1 /* same issue with Trait2 here */>::foo("fo"); // Choose declaration: foo or foo
}
```

**Comment by sollyucko** [comment]
https://github.com/intellij-rust/intellij-rust/commit/01cf7046cc74850a546256e71f5a6c13e2858faa seems to have fixed most of this. It still often has duplicate results, however, such as if `u8` is changed to `[u8; 1]` in my first example:

```rust
use serde_hex::{SerHex, Strict};
use serde_json::Serializer;

fn foo() {
    <[u8; 1] as SerHex<Strict>>::serialize(&0_u8, Vec::<u8>::new());
}
```

Testing with the revision from `master` and resolving references for `serialize` results in 4 results pointing to the exact same location, the method on the (correct) trait.

**Comment by Kobzol** [comment]
@vlad20012 this issue is caused by the same thing as in #3436 - by this line: https://github.com/intellij-rust/intellij-rust/pull/5661/files#diff-a2b72fdf4730d3704af99c0f7c4d6103L1221. It seems that filtering/sorting inherent vs trait methods has to be done in a later/upper stage of name resolution. Is it worth to change this now or will all of this be changed with name resolution 2.0?

**Comment by vlad20012** [comment]
@Kobzol 
Name Resolution 2.0 is not related to the type-dependent name resolution, i.e. traits, impls, methods, etc. So this code will appear as is even after Name Resolution 2.0 implementation. 

There are multiple issues with this code. One of them was tried to be fixed in #4729 (filterIncompatibleImpls), but I don't like that approach a lot because it is too strictly filters incorrect resolve variants.

Another one is an issue that tried to be fixed in #5661 but it brings multiresolve instead of correct resolve to a visible trait item.

I think there are some architectural problems. We should process type-related paths in a way similar to method resolution (see `pickSingleMethod`/`filterAssocItems`), i.e. collect all possible associated items and then select correct ones. I've been trying to implement this for a long time, but more priority/interesting tasks always appearing. I hope I'll fix it soon =)
